File: Source/RTSPPlayerServer/RTSPPlayerServer/Utilities/ChronoUtilities.cs
using System;

namespace RTSPPlayerServer.Utilities
{
    /// <summary>
    /// An utility class that provides methods for calculating timestamps.
    /// </summary>
    internal static class ChronoUtilities
    {
        /// <summary>
        /// Latest timestamp in microseconds.
        /// </summary>
        private static long _baseMicroseconds = DateTime.Now.Ticks / 10L;

        /// <summary>
        /// Generates a 32-bit timestamp from microseconds.
        /// </summary>
        /// <returns>A 32-bit timestamp from microseconds.</returns>
        public static uint TimestampMicroseconds32()
        {
            var currentMicroseconds = DateTime.Now.Ticks / 10L;

            _baseMicroseconds = currentMicroseconds <= _baseMicroseconds
                ? _baseMicroseconds + 1
                : currentMicroseconds;

            return (uint) _baseMicroseconds;
        }
    }
}

File: Source/RTSPPlayerClient/Common/Utility/ChronoUtilities.cpp
/// \file ChronoUtilities.cpp
/// \brief Contains definitions of utility classes and functions for calculating
/// timestamps.
/// \bug No known bugs.

#include "ChronoUtilities.hpp"

#include <chrono>

/// A namespace that contains common utility classes and functions.
namespace Common::Utility {

    /// Generates a 64-bit timestamp from microseconds.
    /// \details Generates a 64-bit timestamp using steady clock.
    /// \return A 64-bit timestamp from microseconds.
    quint64 timestampMicroseconds64() noexcept {
        static auto base =
            std::chrono::time_point_cast<std::chrono::microseconds>(
                std::chrono::steady_clock::now());

        auto current =
            std::chrono::time_point_cast<std::chrono::microseconds>(
                std::chrono::steady_clock::now());

        if (current <= base) base += std::chrono::microseconds(1);
        else base = current;

        return base.time_since_epoch().count();
    }

    /// Generates a 32-bit timestamp from microseconds.
    /// \details Generates a 32-bit timestamp using steady clock.
    /// \return A 32-bit timestamp from microseconds.
    quint32 timestampMicroseconds32() noexcept {
        return static_cast<quint32>(timestampMicroseconds64());
    }
}
